### PR TITLE
Adopt lagom keystore generation changes in Play

### PR DESCRIPTION
**Pull request by marcospereira**
## Purpose

Lagom has a copy of our local certificate generation, but with some adjustments. See https://github.com/lagom/lagom/pull/1443.

Refs https://github.com/lagom/lagom/issues/1507

This brings most of the changes made there so that both frameworks can use the same code.

## Missing tasks

- [x] Documentation (migration guide)

### CODE REVIEW

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala [hunk 1]
```diff
@@ -5,60 +5,87 @@
...
     val key: RSAPublicKey = c.getPublicKey.asInstanceOf[RSAPublicKey]
     key.getModulus.bitLength < 2048 || c.asInstanceOf[X509CertImpl].getSigAlgName != SignatureAlgorithmName
   }
 
-  def keyManagerFactory(appPath: File): KeyManagerFactory = {
+  def createKeyStore(appPath: File): KeyStore = {
     val keyStoreFile = new File(appPath, GeneratedKeyStore)
+
+    // Ensure directory for keystore exists
+    keyStoreFile.getParentFile.mkdirs()
```
Comment (wsargent) [inline framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: should check for return value here

Comment (marcospereira) [reply framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: Done. See `createKeystoreParentDirectory` method. Should better handle the cases where the dir was not created.

Thanks for reviewing, @wsargent.

File: documentation/manual/working/commonGuide/production/ConfiguringHttps.md [hunk 2]
```diff
@@ -26,30 +28,31 @@ Having created your keystore, the following configuration properties can be used
...
+sbt -Dhttps.port=9443 -Dplay.server.https.engineProvider=mypackage.CustomSSLEngineProvider run
+```
 
 ## Turning HTTP off
 
 To disable binding on the HTTP port, set the `http.port` system property to be `disabled`, eg:
 
-    ./start -Dhttp.port=disabled -Dhttps.port=9443 -Dplay.server.https.keyStore.path=/path/to/keystore -Dplay.server.https.keyStore.password=changeme
+```bash
+sbt -Dhttp.port=disabled -Dhttps.port=9443 -Dplay.server.https.keyStore.path=/path/to/keystore -Dplay.server.https.keyStore.password=changeme run
```
Comment (renatocaval) [inline documentation/manual/working/commonGuide/production/ConfiguringHttps.md]: Here we are also configuring the key store path and password, but the main topic of that section is about disabling http. 

I would only add:
```
./start -Dhttp.port=disabled -Dhttps.port=9443
```

File: documentation/manual/working/commonGuide/production/ConfiguringHttps.md [hunk 3]
```diff
@@ -3,7 +3,9 @@
 
 Play can be configured to serve HTTPS.  To enable this, simply tell Play which port to listen to using the `https.port` system property.  For example:
 
-    ./start -Dhttps.port=9443
+```bash
+sbt -Dhttps.port=9443 run
```
Comment (renatocaval) [inline documentation/manual/working/commonGuide/production/ConfiguringHttps.md]: why switching from `./start` to `sbt` command? Isn't that part of the documentation explaining how to do it in prod?

Is that consistent with the rest of the docs? 
I mean, which flavour do we favour? Dev mode or prod mode?

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala [hunk 4]
```diff
@@ -5,60 +5,91 @@
...
+
+  def fileInDevModeDir(filename: String): String = {
+    "target" + File.separatorChar + "dev-mode" + File.separatorChar + filename
+  }
 
-  def shouldGenerate(keyStoreFile: File): Boolean = {
+  /**
+   * @param appPath a file descriptor to the root folder of the project (the root, not a particular module).
+   */
+  def getKeyStoreFilePath(appPath: File) = new File(appPath, GeneratedKeyStore)
```
Comment (renatocaval) [inline framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: this method seems not to be used.

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala [hunk 5]
```diff
@@ -96,18 +151,59 @@ object FakeKeyStore {
...
+  def createSelfSignedCertificate(keyPair: KeyPair, certificateAuthorityKeyPair: KeyPair): X509Certificate = {
+    val certInfo = new X509CertInfo()
+
+    // Serial number and version
+    certInfo.set(X509CertInfo.SERIAL_NUMBER, new CertificateSerialNumber(new BigInteger(64, new SecureRandom())))
+    certInfo.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3))
+
+    // Validity
+    val validFrom = new Date()
+    val validTo = new Date(validFrom.getTime + 50l * 365l * 24l * 60l * 60l * 1000l)
```
Comment (renatocaval) [inline framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: 50 years! 😱

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala [hunk 6]
```diff
@@ -130,6 +226,11 @@ object FakeKeyStore {
     val algorithm = new AlgorithmId(SignatureAlgorithmOID)
     certInfo.set(X509CertInfo.ALGORITHM_ID, new CertificateAlgorithmId(algorithm))
 
+    val caExtension =
+      new CertificateExtensions
+    caExtension.set(BasicConstraintsExtension.NAME, new BasicConstraintsExtension( /* isCritical */ true, /* isCA */ true, 0))
```
Comment (renatocaval) [inline framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: names parameters the Java Way™ 👍

File: documentation/manual/releases/release27/migration27/Migration27.md [hunk 7]
```diff
@@ -369,3 +369,7 @@ CoordinatedShutdown.Reason reason = CoordinatedShutdown.unknownReason();
 Optional<String> runFromPhase = Optional.of("");
 CoordinatedShutdown.get(actorSystem).run(reason, runFromPhase);
 ```
+
+## Change in self-signed HTTPS certificate
+
+It is now generated under `target/dev-mode/generated.keystore` instead of directly on the root folder.
```
Comment (renatocaval) [inline documentation/manual/releases/release27/migration27/Migration27.md]: I would prefer to have it generated under ~/.play, so users don't have to add an exception on their browser over and over again. 

Anyway, that can be the subject of another PR.

File: documentation/manual/working/commonGuide/production/ConfiguringHttps.md (hunk 2 above)
Comment (marcospereira) [reply documentation/manual/working/commonGuide/production/ConfiguringHttps.md]: Yeah, it is just that `start` does not exists anymore. It was legacy stuff. But you are correct that we want to have examples for production usage.

I will update accordingly.

File: documentation/manual/working/commonGuide/production/ConfiguringHttps.md (hunk 3 above)
Comment (marcospereira) [reply documentation/manual/working/commonGuide/production/ConfiguringHttps.md]: Yeah, it is just that `start` does not exists anymore. It was legacy stuff. But you are correct that we want to have examples for production usage.

I will update accordingly.

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala (hunk 4 above)
Comment (marcospereira) [reply framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: Good catch!

It should be used inside of `createKeyStore`.

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala (hunk 6 above)
Comment (marcospereira) [reply framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: 🙃

File: documentation/manual/releases/release27/migration27/Migration27.md (hunk 7 above)
Comment (marcospereira) [reply documentation/manual/releases/release27/migration27/Migration27.md]: Yes, I thought about something similar (keep using project root) but decided to leave it to another PR.

The ultimate thing here would be to install the CA automatically.

**Comment by marcospereira** [comment]
@renatocaval, @richdougherty, and @wsargent I think this is ready to be merged.

Appreciate if you can do a final review here.

Best.

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala [hunk 8]
```diff
@@ -5,60 +5,91 @@
...
 
-  def certificateTooWeak(c: java.security.cert.Certificate): Boolean = {
+  private[play] def certificateTooWeak(c: java.security.cert.Certificate): Boolean = {
     val key: RSAPublicKey = c.getPublicKey.asInstanceOf[RSAPublicKey]
     key.getModulus.bitLength < 2048 || c.asInstanceOf[X509CertImpl].getSigAlgName != SignatureAlgorithmName
   }
 
-  def keyManagerFactory(appPath: File): KeyManagerFactory = {
-    val keyStoreFile = new File(appPath, GeneratedKeyStore)
+  def createKeyStore(appPath: File): KeyStore = {
```
[the remark this answers is not in the document]
Comment (marcospereira) [reply framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: Needs to be public for Lagom.

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala [hunk 9]
```diff
@@ -5,60 +5,91 @@
...
+    val keyStoreFile = getKeyStoreFilePath(appPath)
```
[the remark this answers is not in the document]
Comment (marcospereira) [reply framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: It is just that it has a different name now.

File: framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (marcospereira) [reply framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: Also have a different name now.

[the remark this answers is not in the document]
Comment (TimMoore) [reply framework/src/play-server/src/main/scala/play/core/server/ssl/FakeKeyStore.scala]: In the future, it might be a good idea to comment on APIs that are exposed publicly specifically for Lagom.
